Sources/NIOHTTP2/HTTP2PipelineHelpers.swift

```diff
@@ -104,4 +105,86 @@ extension Channel {
+    func configureHTTP2SecureUpgrade(h2ChannelConfigurator: @escaping (Channel) -> EventLoopFuture<Void>,
+                                     http1ChannelConfigurator: @escaping (Channel) -> EventLoopFuture<Void>) -> EventLoopFuture<Void> {
+        let alpnHandler = ApplicationProtocolNegotiationHandler { result in
+            switch result {
+            case .negotiated("h2"):
+                // Successful upgrade to HTTP/2. Let the user configure the pipeline.
+                return h2ChannelConfigurator(self)
+            case .negotiated("http/1.1"), .fallback:
+                // Explicit or implicit HTTP/1.1 choice.
+                return http1ChannelConfigurator(self)
+            case .negotiated:
+                // We negotiated something that isn't HTTP/1.1. This is a bad scene, and is a good indication
+                // of a user configuration error. We're going to close the connection directly.
+                return self.close().flatMap { self.eventLoop.makeFailedFuture(NIOHTTP2Errors.InvalidALPNToken()) }
+            }
+        }
+
+        return self.pipeline.addHandler(alpnHandler)
+    }
+
+    /// Configures a `ChannelPipeline` to speak either HTTP or HTTP/2 accordingly to what can be negotiated with the client.
+    ///
+    /// This helper takes care of configuring the server pipeline such that it negotiates whether to
+    /// use HTTP/1.1 or HTTP/2. Once the protocol to use for the channel has been negotitated, the
+    /// provided callback will configure the application-specific handlers in a protocol-agnostic way.
+    ///
+    /// This function does't configure the TLS handler. Callers of this function need to add a TLS
+    /// handler appropriately configured to perform protocol negotiation.
+    ///
+    /// - parameters:
+    ///     - h2ConnectionChannelConfigurator: A callback that will be invoked only
+    ///         when the negotiated protocol is H2 to configure the connection channel.
+    ///     - configurator: A callback that will be invoked after a protocol has been negotiated.
+    ///         The callback only needs to add application-specific handlers and must return a future
+    ///         that completes when the channel has been fully mutated.
+    /// - returns: `EventLoopFuture<Void>` that completes when the channel is ready.
+    public func configureCommonHTTPServerPipeline(
+        h2ConnectionChannelConfigurator: @escaping (Channel) -> EventLoopFuture<Void>,
+        _ configurator: @escaping (Channel) -> EventLoopFuture<Void>) -> EventLoopFuture<Void> {
+        let h2ChannelConfigurator = { (channel: Channel) -> EventLoopFuture<Void> in
+            channel.configureHTTP2Pipeline(mode: .server) { (streamChannel, streamID) -> EventLoopFuture<Void> in
+                streamChannel.pipeline.addHandler(HTTP2ToHTTP1ServerCodec(streamID: streamID)).flatMap { () -> EventLoopFuture<Void> in
+                    configurator(streamChannel)
```

Comment:
@Lukasa is it fine to expect that `configurator` do the channel close I previously added separately with `ErrorHandler`?

The reason I added was a cargo-culting of what I've seen in https://github.com/apple/swift-nio-examples/blob/master/http2-server/Sources/http2-server/main.swift#L137:L139 but after having another look I think we might just want to configure both at the same time in the callback.

Reply:
Yes, that's totally fine.